### DISCUSSION TITLE
Has NetBeans 19 changed the life time of the "mvnd" process?

### DISCUSSION THREAD

**Discussion by negora** [apache/netbeans#6470] [Q&A]
Hi:

I've configured NetBeans to use `mvnd` instead of `mvn`. With NetBeans 18, the server process of `mvnd` is kept running for hours (3 hours maximum being idle; I can see it in the parameters of the command). However, with NetBeans 19, the process finishes after 10 seconds approximately. This makes `mvnd` useless if you don't run it within those 10 seconds. Is it a bug? Or is there something that I'm missing?

I've compared the commands generated by version 18 and 19 and they're exactly the same (except the `mvnd.id` parameter, of course). I don't know if the `netbeans-eventspy.jar` library is doing something different in version 19.

Thank you for your help.

**Comment by neilcsmith-net** [comment]
cc/ @mbien for input.  I think most of those things are inside `mvnd` itself.  I'm not aware we changed anything that would impact this in NB19.  Have you changed the daemon version in use?  What about JDK?  That you said 10 seconds is interesting, as if I understand correctly that's the frequency with which the daemon checks whether to exit.

**Reply by mbien** [reply]
NB doesn't even know that there is a daemon.

see https://github.com/apache/netbeans/issues/6472#issuecomment-1728227583 or update to latest mvnd.

**Comment by negora** [comment]
@neilcsmith-net I'm using exactly the same versions and installations of JDK and `mvnd` with both versions of NetBeans:

* JDK 17.0.7+7-Debian-1deb11u1.
* mvnd 1.0-m6-m39.

I've made a another test with version 19, starting from a clean configuration, and I think that I've just found the origin of the problem. I've set `mvnd` as the command, and the process has survived beyond 10 seconds. But as soon as I've set the parameter `--threads 1`, the new process closes after 10 seconds. I don't understand why version 18 works as expected but 19 doesn't. After all, this parameter only affects `mvnd`, right?

Just in case you wonder why I limit it to a single thread, it's because the output is easier to read to me.

PS: if I set `--threads 2`, the process doesn't close prematurely. So the cause is the fact of setting 1 thread, not the fact of setting any quantity of threads.

**Reply by mbien** [reply]
serial/single threaded modes will enable slightly more output window features like folding for example (and register the eventspy) which might influence the timing. 

mvn is sequential by default while mvnd is parallel by default, so NB is parsing the commands and figuring out what the output will be when combined with the mvn/mvnd executable.

**Comment by negora** [comment]
@neilcsmith-net I'm note sure if this is a clue in the right direction or has nothing to do. If I don't pass the `--threads` parameter to `mvnd`, or if pass `--threads 2`, NetBeans calls the command without the `-Dmvnd.extClasspath` parameter. However, if I set `--threads 1`, NetBeans passes the following parameter:

```
-Dmvnd.extClasspath=/usr/local/lib/netbeans/java/maven-nblib/netbeans-eventspy.jar
```

But this behaviour is the same in versions 18 and 19... 

PS: I believe that NetBeans uses an environment variable to set `-Dmvnd.extClasspath`. Because I've tried to manually run the commands that NetBeans shows in the *Output* panel (with and without thread configuration), and `mvnd` behaves well and the processes don't close before time.

**Reply by neilcsmith-net** [reply]
OK.  Out of interest, does `--serial` have the same problem?

Please open an issue on this one - looks like something that might need investigating leading up to NB20.

**Reply by negora** [reply]
> OK. Out of interest, does `--serial` have the same problem?

Confirmed: `--serial` causes the same issue.

> Please open an issue on this one - looks like something that might need investigating leading up to NB20.

OK. Give me some time to finish some tasks and I'll open it.

By the way, thank you for your help!

**Reply by negora** [reply]
OK, I've created the issue #6472 .
